File: StreamDVR/StreamDVR/StreamRecorder.swift
import Foundation

class StreamRecorder {
    private var process: Process?
    private var completionHandler: ((String) -> Void)?
    private var isRunning = false
    private var fileHandle: FileHandle?
    private(set) var outputPath: String?

    func startRecording(url: String, outputDir: String, channelName: String, accessToken: String?, platform: StreamPlatform, streamInfo: StreamInfo, completion: @escaping (String) -> Void) {
        self.completionHandler = completion
        let streamlinkPath = findStreamlink()
        guard let path = streamlinkPath else {
            completion("Error: streamlink not found. Install with: brew install streamlink")
            return
        }

        let now = Date()
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        let dateString = formatter.string(from: now)

        formatter.dateFormat = "HH-mm-ss"
        let timeString = formatter.string(from: now)

        let safeChannel = Self.sanitizeFilename(channelName)
        let finalChannel = safeChannel.isEmpty ? "channel" : safeChannel

        let safeTitle = Self.sanitizeFilename(streamInfo.title)
        let finalTitle = safeTitle.isEmpty ? "stream" : String(safeTitle.prefix(35))

        let filename = "\(finalChannel)_\(dateString)-\(timeString)_\(finalTitle).ts"
        let outputPath = (outputDir as NSString).appendingPathComponent(filename)
        try? FileManager.default.createDirectory(atPath: outputDir, withIntermediateDirectories: true)
        self.outputPath = outputPath

        var args: [String] = []

        // Twitch-only flags: ad removal + premium ad-free auth via the web token.
        if platform == .twitch {
            args.append(contentsOf: ["--twitch-disable-ads"])

            // The web auth-token (from the login cookie) authenticates Twitch API calls
            // so premium / ad-free users get a clean stream. Sent as an API header.
            // Note: streamlink expects KEY=VALUE format; streamlink 8.x no longer has --twitch-client-id.
            if let token = accessToken, !token.isEmpty {
                args.append(contentsOf: ["--twitch-api-header", "Authorization=OAuth \(token)"])
            }
        }

        // Kick: optional session cookies are forwarded to streamlink so requests
        // are made as the signed-in user (recording still works without them).
        if platform == .kick {
            for (name, value) in KickSession.cookies {
                args.append(contentsOf: ["--http-cookie", "\(name)=\(value)"])
            }
        }

        args.append(contentsOf: [
            "-o", outputPath,
            "--force",
            "--stream-segment-attempts", "10",
            "--stream-segment-timeout", "30",
            "--retry-open", "30",
            "--retry-streams", "10",
            "--retry-max", "10",
            "--ringbuffer-size", "32M"
        ])

        args.append(url)
        args.append("best")

        let proc = Process()
        proc.executableURL = URL(fileURLWithPath: path)
        proc.arguments = args

        proc.environment = Self.augmentedEnvironment()

        let pipe = Pipe()
        proc.standardOutput = pipe
        proc.standardError = pipe

        pipe.fileHandleForReading.readabilityHandler = { handle in
            // Drain stdout/stderr so streamlink never blocks on a full pipe.
            // Errors surface via the process exit status and finalization log.
            _ = handle.availableData
        }

        proc.terminationHandler = { [weak self] proc in
            DispatchQueue.main.async {
                self?.isRunning = false
                pipe.fileHandleForReading.readabilityHandler = nil
                if proc.terminationStatus == 0 || proc.terminationStatus == 1 {
                    self?.complete(outputPath)
                } else {
                    self?.complete("Recording stopped (exit code \(proc.terminationStatus))")
                }
            }
        }

        do {
            try proc.run()
            self.process = proc
            self.isRunning = true
        } catch {
            completion("Failed to start streamlink: \(error.localizedDescription)")
        }
    }

    func stop() {
        if let proc = process, proc.isRunning {
            proc.terminate()
            DispatchQueue.main.asyncAfter(deadline: .now() + 2) { [weak proc] in
                if let p = proc, p.isRunning {
                    p.interrupt()
                }
            }
        }
        isRunning = false
    }

    /// Finalizes the recording: if the output file was created but no real data
    /// ever arrived (some streams, e.g. very low bitrate starts, produce an
    /// empty/header-only .ts), it is discarded instead of being left behind.
    /// Otherwise the file's base PTS is normalized to ~0 so players start at
    /// 00:00 (some sources keep a large base timestamp, e.g. ~1h44m).
    private func complete(_ output: String) {
        guard let completion = completionHandler else { return }
        let stoppedOrFailed = output.hasPrefix("Recording stopped") || output.hasPrefix("Error")
        guard let path = self.outputPath else {
            completion(output)
            return
        }
        if stoppedOrFailed, !FileManager.default.fileExists(atPath: path) {
            completion(output)
            return
        }
        let size = (try? FileManager.default.attributesOfItem(atPath: path)[.size] as? Int64) ?? -1
        if FileManager.default.fileExists(atPath: path), size < 1024 {
            try? FileManager.default.removeItem(atPath: path)
            completion("discarded:\(path)")
            return
        }
        DispatchQueue.global(qos: .utility).async {
            let result = Self.finalizeToMp4(path, fallbackOutput: output)
            DispatchQueue.main.async {
                completion(result)
            }
        }
    }

    /// Converts a finished recording to MP4 with stream copy (fast, no
    /// re-encoding): MP4 rebases timestamps so every player starts at 00:00.
    /// Falls back to re-encoding only the audio if the copy is refused
    /// (e.g. opus audio on LL-HLS), and keeps the original .ts on failure.
    private static func finalizeToMp4(_ path: String, fallbackOutput: String) -> String {
        guard path.hasSuffix(".ts") else { return fallbackOutput }
        let mp4 = String(path.dropLast(3)) + ".mp4"

        Self.runProcess("ffmpeg", ["-y", "-v", "error", "-i", path, "-map", "0", "-c", "copy", "-movflags", "+faststart", mp4])
        if FileManager.default.fileExists(atPath: mp4),
           let size = try? FileManager.default.attributesOfItem(atPath: mp4)[.size] as? Int64,
           size > 0 {
            try? FileManager.default.removeItem(atPath: path)
            return "converted:\(mp4)"
        }
        try? FileManager.default.removeItem(atPath: mp4)

        // Fallback: transcode audio only (video stays copied).
        Self.runProcess("ffmpeg", ["-y", "-v", "error", "-i", path, "-map", "0", "-c:v", "copy", "-c:a", "aac", "-movflags", "+faststart", mp4])
        if FileManager.default.fileExists(atPath: mp4),
           let size = try? FileManager.default.attributesOfItem(atPath: mp4)[.size] as? Int64,
           size > 0 {
            try? FileManager.default.removeItem(atPath: path)
            return "converted:\(mp4)"
        }
        try? FileManager.default.removeItem(atPath: mp4)
        return "convertfailed:\(path)"
    }

    /// Runs a tool found on PATH (ffprobe/ffmpeg from the Homebrew tool dirs)
    /// and returns its captured stdout, or nil on failure.
    @discardableResult
    private static func runProcess(_ tool: String, _ arguments: [String]) -> String? {
        let proc = Process()
        proc.executableURL = URL(fileURLWithPath: "/usr/bin/env")
        proc.arguments = [tool] + arguments
        proc.environment = augmentedEnvironment()
        let pipe = Pipe()
        proc.standardOutput = pipe
        proc.standardError = FileHandle.nullDevice
        do {
            try proc.run()
        } catch {
            return nil
        }
        let data = pipe.fileHandleForReading.readDataToEndOfFile()
        proc.waitUntilExit()
        return String(data: data, encoding: .utf8)
    }

    /// GUI-launched apps have a minimal PATH without the Homebrew tool dirs.
    /// streamlink needs ffmpeg on PATH to use its ffmpeg muxer; without it,
    /// Chaturbate (LL-HLS/fMP4) recordings lose audio and keep the source
    /// stream's large base PTS (a long blank start). Expose the tool dirs.
    private static func augmentedEnvironment() -> [String: String] {
        var environment = ProcessInfo.processInfo.environment
        let toolBins = ["/opt/homebrew/bin", "/usr/local/bin"]
        if let currentPath = environment["PATH"] {
            environment["PATH"] = (toolBins + [currentPath]).joined(separator: ":")
        } else {
            environment["PATH"] = (toolBins + ["/usr/bin", "/bin", "/usr/sbin", "/sbin"]).joined(separator: ":")
        }
        return environment
    }

    static func sanitizeFilename(_ input: String) -> String {
        input
            .replacingOccurrences(of: "/", with: "-")
            .replacingOccurrences(of: ":", with: "-")
            .replacingOccurrences(of: "\\", with: "-")
            .replacingOccurrences(of: "\"", with: "")
            .replacingOccurrences(of: "*", with: "")
            .replacingOccurrences(of: "?", with: "")
            .replacingOccurrences(of: "<", with: "")
            .replacingOccurrences(of: ">", with: "")
            .replacingOccurrences(of: "|", with: "-")
            .trimmingCharacters(in: .whitespaces)
    }

    private func findStreamlink() -> String? {
        let paths = [
            "/opt/homebrew/bin/streamlink",
            "/usr/local/bin/streamlink",
            "/usr/bin/streamlink",
            "/opt/local/bin/streamlink",
            "/usr/sbin/streamlink"
        ]

        for path in paths {
            if FileManager.default.fileExists(atPath: path) {
                return path
            }
        }

        let task = Process()
        task.executableURL = URL(fileURLWithPath: "/bin/bash")
        task.arguments = ["-c", "which streamlink 2>/dev/null || echo ''"]
        let pipe = Pipe()
        task.standardOutput = pipe
        task.standardError = FileHandle.nullDevice
        try? task.run()
        task.waitUntilExit()

        let data = pipe.fileHandleForReading.readDataToEndOfFile()
        let path = String(data: data, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines)
        if let path = path, !path.isEmpty, FileManager.default.fileExists(atPath: path) {
            return path
        }

        return nil
    }
}

File: Windows/StreamDVR.Core/StreamRecorder.cs
using System.Diagnostics;

namespace StreamDVR.Core;

public class StreamRecorderResult
{
    public bool Started { get; set; }
    public string OutputPath { get; set; } = "";
    public string Message { get; set; } = "";
    public Process? Process { get; set; }
}

/// <summary>
/// Wraps the streamlink process (Windows edition). Finds streamlink/ffprobe
/// on PATH and in the common Streamlink install locations.
/// </summary>
public static class StreamRecorder
{
    private static readonly string[] StreamlinkLocations =
    {
        Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "Programs", "streamlink", "bin", "streamlink.exe"),
        @"C:\Program Files\Streamlink\bin\streamlink.exe",
        @"C:\Program Files (x86)\Streamlink\bin\streamlink.exe",
        Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "Programs", "Streamlink", "bin", "streamlink.exe")
    };

    private static readonly string[] FfprobeLocations =
    {
        Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "Programs", "ffmpeg", "bin", "ffprobe.exe"),
        @"C:\Program Files\ffmpeg\bin\ffprobe.exe",
        @"C:\ffmpeg\bin\ffprobe.exe",
        Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "Microsoft", "WinGet", "Packages", "Gyan.FFmpeg", @"ffmpeg-*-build\bin\ffprobe.exe")
    };

    public static string? FindStreamlink()
    {
        foreach (var loc in StreamlinkLocations)
        {
            var expanded = ExpandWildcards(loc);
            if (File.Exists(expanded)) return expanded;
        }
        return FindOnPath("streamlink");
    }

    public static string? FindFfprobe()
    {
        foreach (var loc in FfprobeLocations)
        {
            if (File.Exists(loc)) return loc;
        }
        return FindOnPath("ffprobe");
    }

    /// ffmpeg normally sits in the same bin dir as ffprobe (winget Gyan.FFmpeg).
    public static string? FindFfmpeg()
    {
        foreach (var loc in FfprobeLocations)
        {
            var dir = Path.GetDirectoryName(ExpandWildcards(loc));
            if (dir == null) continue;
            var exe = Path.Combine(dir, "ffmpeg.exe");
            if (File.Exists(exe)) return exe;
        }
        return FindOnPath("ffmpeg");
    }

    /// If the recording's start timestamp is far from zero, remuxes it with
    /// stream copy (fast, lossless) so players begin at 00:00 instead of, e.g.,
    /// ~1h44m (the source stream keeps a large base PTS for LL-HLS/fMP4).
    public static string? ConvertToMp4(string path)
    {
        try
        {
            var info = new FileInfo(path);
            if (!info.Exists || info.Length < 1024) return null;
            if (!path.EndsWith(".ts", StringComparison.OrdinalIgnoreCase)) return null;

            var ffmpeg = FindFfmpeg();
            if (ffmpeg == null) return "error:ffmpeg not found";

            var mp4 = path[..^3] + ".mp4";

            // Preferred: stream copy (fast, no re-encoding); MP4 rebases
            // timestamps so players start at 00:00.
            RunTool(ffmpeg, "-y", "-v", "error", "-i", path, "-map", "0", "-c", "copy", "-movflags", "+faststart", mp4);
            if (File.Exists(mp4) && new FileInfo(mp4).Length > 0)
            {
                File.Delete(path);
                return $"ok:{mp4}";
            }
            try { File.Delete(mp4); } catch { }

            // Fallback: transcode audio only (video stays copied).
            RunTool(ffmpeg, "-y", "-v", "error", "-i", path, "-map", "0", "-c:v", "copy", "-c:a", "aac", "-movflags", "+faststart", mp4);
            if (File.Exists(mp4) && new FileInfo(mp4).Length > 0)
            {
                File.Delete(path);
                return $"ok:{mp4}";
            }
            try { File.Delete(mp4); } catch { }
            return "error:MP4 conversion failed";
        }
        catch { return null; }
    }

    private static string? RunTool(string exe, params string[] args)
    {
        try
        {
            var psi = new ProcessStartInfo(exe)
            {
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true
            };
            foreach (var a in args) psi.ArgumentList.Add(a);
            using var proc = Process.Start(psi);
            if (proc == null) return null;
            var outText = proc.StandardOutput.ReadToEnd();
            proc.WaitForExit(600_000);
            return outText;
        }
        catch { return null; }
    }

    private static string ExpandWildcards(string path)
    {
        if (!path.Contains('*')) return path;
        var dir = Path.GetDirectoryName(path);
        var pattern = Path.GetFileName(path);
        if (dir == null || !Directory.Exists(dir)) return path;
        var match = Directory.EnumerateDirectories(dir, pattern).FirstOrDefault();
        return match ?? path;
    }

    private static string? FindOnPath(string name)
    {
        try
        {
            var psi = new ProcessStartInfo("where.exe", name)
            {
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true
            };
            using var proc = Process.Start(psi);
            if (proc == null) return null;
            var outPath = proc.StandardOutput.ReadToEnd().Trim();
            proc.WaitForExit(5000);
            var first = outPath.Split('\n', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries).FirstOrDefault();
            return File.Exists(first) ? first : null;
        }
        catch
        {
            return null;
        }
    }

    public static StreamRecorderResult Start(string url, string outputDir, string channelName, string? accessToken,
        StreamPlatform platform, StreamInfo streamInfo)
    {
        var streamlink = FindStreamlink();
        if (streamlink == null)
        {
            return new StreamRecorderResult
            {
                Message = "Error: streamlink not found. Install with: winget install streamlink"
            };
        }

        var now = DateTime.Now;
        var dateString = now.ToString("yyyy-MM-dd");
        var timeString = now.ToString("HH-mm-ss");

        var safeChannel = SanitizeFilename(channelName);
        var finalChannel = string.IsNullOrEmpty(safeChannel) ? "channel" : safeChannel;

        var safeTitle = SanitizeFilename(streamInfo.Title);
        var finalTitle = string.IsNullOrEmpty(safeTitle) ? "stream" : CapLength(safeTitle, 35);
        var filename = $"{finalChannel}_{dateString}-{timeString}_{finalTitle}.ts";
        var outputPath = Path.Combine(outputDir, filename);

        var args = new List<string>();

        // Twitch-only flags: ad removal + premium ad-free auth via the web token.
        if (platform == StreamPlatform.Twitch)
        {
            args.Add("--twitch-disable-ads");
            if (!string.IsNullOrEmpty(accessToken))
            {
                args.Add("--twitch-api-header");
                args.Add($"Authorization=OAuth {accessToken}");
            }
        }

        // Kick: optional session cookies are forwarded to streamlink so requests are
        // made as the signed-in user (recording still works without them).
        if (platform == StreamPlatform.Kick)
        {
            foreach (var cookie in KickSession.Cookies)
            {
                args.Add("--http-cookie");
                args.Add($"{cookie.Key}={cookie.Value}");
            }
        }

        args.AddRange(new[]
        {
            "-o", outputPath,
            "--force",
            "--stream-segment-attempts", "10",
            "--stream-segment-timeout", "30",
            "--retry-open", "30",
            "--retry-streams", "10",
            "--retry-max", "10",
            "--ringbuffer-size", "32M",
            url,
            "best"
        });

        try
        {
            Directory.CreateDirectory(outputDir);
            var psi = new ProcessStartInfo
            {
                FileName = streamlink,
                UseShellExecute = false,
                CreateNoWindow = true,
                RedirectStandardOutput = true,
                RedirectStandardError = true
            };
            foreach (var arg in args)
            {
                psi.ArgumentList.Add(arg);
            }

            var proc = Process.Start(psi);
            if (proc == null)
            {
                return new StreamRecorderResult { Message = "Failed to start streamlink" };
            }

            proc.OutputDataReceived += (_, e) => { OnOutput(e.Data); };
            proc.ErrorDataReceived += (_, e) => { OnOutput(e.Data); };
            proc.BeginOutputReadLine();
            proc.BeginErrorReadLine();

            return new StreamRecorderResult
            {
                Started = true,
                OutputPath = outputPath,
                Process = proc,
                Message = $"Recording {finalTitle}"
            };
        }
        catch (Exception ex)
        {
            return new StreamRecorderResult { Message = $"Failed to start streamlink: {ex.Message}" };
        }
    }

    public static void Stop(Process? proc)
    {
        if (proc == null || proc.HasExited) return;
        try { proc.Kill(entireProcessTree: true); } catch { }
    }

    private static void OnOutput(string? data)
    {
        // Intentionally unused for now; UI reads file + ffprobe for stats.
    }

    private static string SanitizeFilename(string title)
    {
        var invalid = Path.GetInvalidFileNameChars();
        var cleaned = new string(title.Select(c => invalid.Contains(c) ? '-' : c).ToArray());
        return cleaned.Trim().TrimEnd('.');
    }

    /// Caps a string to max characters at a grapheme (unicode text element)
    /// boundary so multi-byte characters are never split.
    private static string CapLength(string s, int max)
    {
        if (string.IsNullOrEmpty(s) || s.Length <= max) return s;
        var result = new System.Text.StringBuilder();
        var enumerator = System.Globalization.StringInfo.GetTextElementEnumerator(s);
        while (enumerator.MoveNext() && result.Length < max)
        {
            result.Append(enumerator.GetTextElement());
        }
        return result.ToString().Trim();
    }

    /// Removes a leftover output file that never received real data
    /// (some streams produce a 0-byte / header-only .ts). Returns true if removed.
    public static bool DeleteIfEmpty(string path)
    {
        try
        {
            var fi = new FileInfo(path);
            if (!fi.Exists) return false;
            if (fi.Length >= 1024) return false;
            fi.Delete();
            return true;
        }
        catch
        {
            return false;
        }
    }
}